File: tests/oracle/fspec_phase0_1204.cc
/*
 * FSPEC-PHASE0-1204: locked Ghidra 12.0.4 oracle for the fspec Phase-0
 * alignment fixes (FSPEC-SPACEFILTER-0002 + FSPEC-TRIALCMP-0003):
 *
 *  - ParamListStandard::findEntry (fspec.cc:661-680) resolves through the
 *    per-space resolverMap: only entries whose space equals the query
 *    address's space are visited. Register/stack entries hit normally;
 *    const/unique-space queries never falsely hit register/ram offsets.
 *  - ParamListStandard::unjustifiedContainer (fspec.cc:1411-1424) and
 *    assumedExtension (fspec.cc:1426-1437) iterate ALL entries with no
 *    space filter (per-entry justifiedContain rejects foreign spaces).
 *  - ParamActive::sortTrials (fspec.hh:316) orders by
 *    ParamTrial::operator< (fspec.cc:1893-1914): model group, then entry
 *    order, then (exclusion) justified offset / (non-exclusion)
 *    reverseStack-aware address, then size; entry-less trials sort last.
 *  - ParamTrial::splitHi/splitLo (fspec.cc:1845/1856): the low piece
 *    starts at addr + (size - sz), both pieces inherit flags;
 *    ParamActive::splitTrial (fspec.cc:2033) renumbers survivor slots and
 *    bumps slotbase.
 *
 * All fixture spaces are little-endian (Rugra's coarse AddressSpace enum
 * defaults to little-endian), and exclusion entries set force_left_justify
 * so the sub-range justification arithmetic is endianness-independent on
 * both sides (the legacy spaceless Address cannot carry isBigEndian —
 * ADDRESS-0001 transitional). Slot observations are deltas because Ghidra
 * slots are 1-based (fspec.cc:4062 `triallist[trial.getSlot()-1]`) while
 * Rugra's coupled consumers are 0-based; the absolute base is outside this
 * fixture's covered projection.
 *
 * ParamEntry/ParamListStandard construction and populateResolver use the
 * same class->struct access hack as the address_space_phase1 /
 * space_registry fixtures; every observation goes through public members
 * only.
 */

#include <bits/stdc++.h>

#define class struct
#define private public
#define protected public
#include "fspec.hh"
#include "opcodes.hh"
#include "space.hh"
#include "translate.hh"
#undef protected
#undef private
#undef class

using namespace ghidra;

class FixtureTranslate final : public Translate {
public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
  }

  std::string tryInsertSpace(AddrSpace *spc) {
    try {
      insertSpace(spc);
      return "ok";
    }
    catch (LowlevelError &err) {
      return std::string("err ") + err.explain;
    }
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    static VarnodeData dummy;
    return dummy;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

// canonical synthetic spaces: constant=0, other=1, unique=2, ram=3,
// register=4, stack=5 (same index plan as the address_space fixtures).
static void buildSpaces(FixtureTranslate *tr,
                        AddrSpace *&constant, AddrSpace *&unique,
                        AddrSpace *&ram, AddrSpace *&reg, AddrSpace *&stack) {
  tr->tryInsertSpace(new ConstantSpace(tr, tr));
  tr->tryInsertSpace(new OtherSpace(tr, tr, OtherSpace::INDEX));
  tr->tryInsertSpace(new UniqueSpace(tr, tr, 2, 0));
  ram = new AddrSpace(tr, tr, IPTR_PROCESSOR, "ram", false, 8, 1, 3,
                      AddrSpace::hasphysical, 0, 0);
  reg = new AddrSpace(tr, tr, IPTR_PROCESSOR, "register", false, 8, 1, 4,
                      AddrSpace::hasphysical, 0, 0);
  stack = new AddrSpace(tr, tr, IPTR_SPACEBASE, "stack", false, 8, 1, 5,
                        AddrSpace::hasphysical, 0, 0);
  tr->tryInsertSpace(ram);
  tr->tryInsertSpace(reg);
  tr->tryInsertSpace(stack);
  constant = tr->getConstantSpace();
  unique = tr->getUniqueSpace();
}

// Fill ParamEntry fields directly (Ghidra fills them in decode(); the
// fixture replicates a staged cspec loader, mirroring Rugra's builder
// setters).
static ParamEntry makeEntry(int4 grp, AddrSpace *spc, uintb base, int4 size,
                            int4 minsize, int4 alignment, uint4 flags) {
  ParamEntry e(grp);
  e.flags = flags;
  e.type = TYPECLASS_GENERAL;
  e.groupSet.clear();
  e.groupSet.push_back(grp);
  e.spaceid = spc;
  e.addressbase = base;
  e.size = size;
  e.minsize = minsize;
  e.alignment = alignment;
  e.numslots = (alignment != 0) ? (size / alignment) : 1;
  e.joinrec = (JoinRecord *)0;
  return e;
}

static std::string loc(AddrSpace *spc, uintb off) {
  std::ostringstream s;
  s << spc->getName() << ":0x" << std::hex << off;
  return s.str();
}

static std::string vd(const VarnodeData &v) {
  std::ostringstream s;
  s << v.space->getName() << ":0x" << std::hex << v.offset << '/' << std::dec
    << v.size;
  return s.str();
}

int main(void) {
  std::cout << "schema=1|fixture=FSPEC-PHASE0-1204|"
               "oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
            << std::endl;

  FixtureTranslate tr;
  AddrSpace *constant, *unique, *ram, *reg, *stack;
  buildSpaces(&tr, constant, unique, ram, reg, stack);

  // entry order fixes both the resolver position order and the
  // operator< entry tie-break.
  ParamListStandard stds;
  // e0: register slot, group 0, exclusion, force-left sub-ranges.
  stds.entry.push_back(makeEntry(0, reg, 0x30, 8, 4, 0,
                                 ParamEntry::force_left_justify));
  // e1: register slot, group 1, exclusion, zero-extends small values.
  stds.entry.push_back(makeEntry(1, reg, 0x38, 8, 4, 0,
                                 ParamEntry::force_left_justify |
                                 ParamEntry::smallsize_zext));
  // e2: stack entry, group 2, aligned slots, reverse stack.
  stds.entry.push_back(makeEntry(2, stack, 0, 64, 4, 8,
                                 ParamEntry::reverse_stack));
  // e3: ram entry, group 3, exclusion, int-typed extension.
  stds.entry.push_back(makeEntry(3, ram, 0x2000, 16, 8, 0,
                                 ParamEntry::force_left_justify |
                                 ParamEntry::smallsize_inttype));
  stds.numgroup = 4;
  stds.populateResolver();

  // ---- case 1: findEntry per-space resolution (via public wrappers) ----
  std::cout << "case=cross_space_find_entry" << std::endl;
  struct PpCase { AddrSpace *spc; uintb off; int4 sz; };
  const PpCase pp[] = {
    { reg, 0x30, 8 },    // justified register hit (group 0)
    { reg, 0x30, 4 },    // force-left sub-justified hit
    { reg, 0x34, 4 },    // unjustified in the entry
    { reg, 0x30, 2 },    // below minsize
    { reg, 0x48, 8 },    // no register entry at this offset
    { stack, 0x8, 8 },   // stack slot 1 (reverse numbering)
    { stack, 0x10, 4 },  // stack slot 2
    { stack, 0x100, 8 }, // beyond the 64-byte stack entry
    { ram, 0x2000, 16 }, // full ram entry hit
    { ram, 0x30, 8 },    // register's offset inside ram: no cross hit
    { unique, 0x30, 8 }, // register's offset in unique: no cross hit
    { constant, 0x2000, 8 }, // ram's offset in const: no cross hit
  };
  for (const PpCase &c : pp) {
    Address a(c.spc, c.off);
    int4 slot = -1, slotsize = -1;
    bool ok = stds.possibleParamWithSlot(a, c.sz, slot, slotsize);
    std::cout << "  pp " << loc(c.spc, c.off) << '/' << c.sz << " ok=" << (ok ? 1 : 0)
              << " slot=" << slot << " slotsize=" << slotsize << std::endl;
  }

  // ---- case 2: unjustifiedContainer has no space filter ----
  std::cout << "case=unjustified_container" << std::endl;
  struct UcCase { AddrSpace *spc; uintb off; int4 sz; };
  const UcCase uc[] = {
    { reg, 0x34, 4 },  // unjustified inside register entry e0
    { reg, 0x32, 4 },  // mid-entry, still unjustified
    { reg, 0x30, 4 },  // justified: must return false
    { ram, 0x2008, 8 }, // unjustified inside ram entry e3
    { stack, 0x2, 8 }, // stack alignment arithmetic (2 % 8 != 0)
  };
  for (const UcCase &c : uc) {
    Address a(c.spc, c.off);
    VarnodeData res;
    bool hit = stds.unjustifiedContainer(a, c.sz, res);
    std::cout << "  uc " << loc(c.spc, c.off) << '/' << c.sz << " hit=" << (hit ? 1 : 0);
    if (hit)
      std::cout << " res=" << vd(res);
    std::cout << std::endl;
  }

  // ---- case 3: assumedExtension has no space filter ----
  std::cout << "case=assumed_extension" << std::endl;
  struct AeCase { AddrSpace *spc; uintb off; int4 sz; };
  const AeCase ae[] = {
    { reg, 0x38, 4 },   // zext flag on the register entry
    { ram, 0x2000, 8 }, // inttype flag on the ram entry
    { reg, 0x30, 4 },   // e0 carries no smallsize flags: COPY
    { ram, 0x2004, 8 }, // unjustified sub-range: COPY
    { reg, 0x38, 8 },   // sz >= entry size: COPY
  };
  for (const AeCase &c : ae) {
    Address a(c.spc, c.off);
    VarnodeData res;
    OpCode op = stds.assumedExtension(a, c.sz, res);
    std::cout << "  ae " << loc(c.spc, c.off) << '/' << c.sz << " op=" << get_opname(op);
    if (op != CPUI_COPY)
      std::cout << " res=" << vd(res);
    std::cout << std::endl;
  }

  // ---- case 4: sortTrials uses operator< (group/entry/offset/addr/size)
  std::cout << "case=sort_trials" << std::endl;
  {
    ParamActive active(false);
    const std::list<ParamEntry> &full = stds.getEntry();
    std::list<ParamEntry>::const_iterator it = full.begin();
    const ParamEntry *elist[4] = { &*it, &*++it, &*++it, &*++it };
    // registered in raw-address order; sorted order must follow the model
    active.registerTrial(Address(stack, 0x10), 8); // stack entry, slot 2
    active.registerTrial(Address(reg, 0x38), 4); // register group 1
    active.registerTrial(Address(stack, 0x0), 8); // stack entry, slot 0
    active.registerTrial(Address(reg, 0x30), 8); // register group 0
    active.registerTrial(Address(reg, 0x34), 4); // register group 0, offset 4
    active.registerTrial(Address(reg, 0x50), 8); // no entry
    active.getTrial(0).setEntry(elist[2], 0);
    active.getTrial(1).setEntry(elist[1], 0);
    active.getTrial(2).setEntry(elist[2], 0);
    active.getTrial(3).setEntry(elist[0], 0);
    active.getTrial(4).setEntry(elist[0], 4);
    active.sortTrials();
    for (int4 i = 0; i < active.getNumTrials(); ++i) {
      const ParamTrial &t = active.getTrial(i);
      const ParamEntry *e = t.getEntry();
      std::cout << "  sorted[" << i << "] addr="
                << loc(t.getAddress().getSpace(), t.getAddress().getOffset())
                << " size=" << t.getSize() << " grp="
                << (e != (const ParamEntry *)0 ? e->getGroup() : -1) << " off="
                << (e != (const ParamEntry *)0 ? t.getOffset() : -1) << std::endl;
    }
  }

  // ---- case 5: splitHi/splitLo address, flags, slot ------------------
  std::cout << "case=split_hi_lo" << std::endl;
  {
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x100), 12);
    ParamTrial &base = active.getTrial(0);
    base.markUsed();
    base.markActive();
    int4 base_slot = base.getSlot();
    ParamTrial hi = base.splitHi(4);
    ParamTrial lo = base.splitLo(4);
    std::cout << "  hi addr=" << loc(reg, hi.getAddress().getOffset())
              << " size=" << hi.getSize()
              << " slot_delta=" << (hi.getSlot() - base_slot)
              << " used=" << (hi.isUsed() ? 1 : 0)
              << " active=" << (hi.isActive() ? 1 : 0)
              << " checked=" << (hi.isChecked() ? 1 : 0) << std::endl;
    std::cout << "  lo addr=" << loc(reg, lo.getAddress().getOffset())
              << " size=" << lo.getSize()
              << " slot_delta=" << (lo.getSlot() - base_slot)
              << " used=" << (lo.isUsed() ? 1 : 0)
              << " active=" << (lo.isActive() ? 1 : 0)
              << " checked=" << (lo.isChecked() ? 1 : 0) << std::endl;
    // the complementary 8-byte low piece
    ParamTrial lo8 = base.splitLo(8);
    std::cout << "  lo8 addr=" << loc(reg, lo8.getAddress().getOffset())
              << " size=" << lo8.getSize()
              << " slot_delta=" << (lo8.getSlot() - base_slot) << std::endl;
  }

  // ---- case 6: splitTrial rebuilds slots and keeps flags --------------
  std::cout << "case=split_trial" << std::endl;
  {
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x100), 12);
    active.registerTrial(Address(reg, 0x200), 8);
    active.getTrial(0).markUsed();
    active.getTrial(0).markActive();
    int4 survivor_slot_before = active.getTrial(1).getSlot();
    int4 slotbase_before = active.slotbase;
    active.splitTrial(0, 4);
    for (int4 i = 0; i < active.getNumTrials(); ++i) {
      const ParamTrial &t = active.getTrial(i);
      std::cout << "  st[" << i << "] addr=" << loc(reg, t.getAddress().getOffset())
                << " size=" << t.getSize() << " used=" << (t.isUsed() ? 1 : 0)
                << " active=" << (t.isActive() ? 1 : 0) << std::endl;
    }
    std::cout << "  survivor_slot_delta="
              << (active.getTrial(2).getSlot() - survivor_slot_before)
              << " slotbase_delta=" << (active.slotbase - slotbase_before)
              << " numtrials=" << active.getNumTrials() << std::endl;
  }

  return 0;
}

File: tests/oracle/fspec_phase0_1204.rs
// FSPEC-PHASE0-1204: Rugra comparand for the locked Ghidra 12.0.4 fspec
// Phase-0 oracle (FSPEC-SPACEFILTER-0002 + FSPEC-TRIALCMP-0003). Mirrors
// tests/oracle/fspec_phase0_1204.cc case for case:
//  - find_entry resolves only entries whose space equals the query space
//    (register/stack entries hit; const/unique/ram cross-offset queries
//    miss),
//  - unjustified_container/assumed_extension iterate every entry with no
//    space filter,
//  - sort_trials orders by ParamTrial::op_less (group, entry order,
//    exclusion offset / reverseStack-aware address, size; entry-less
//    trials last),
//  - split_hi/split_lo produce (addr, size, slot, flags) per fspec.cc:1845/
//    1856 and split_trial renumbers survivor slots and bumps slotbase
//    (fspec.cc:2033).
// Trial addresses are minted through the SpaceRegistry (register/stack) so
// the comparator orders real spaces exactly like Ghidra's tagged Addresses.
// Slot observations are deltas (Ghidra slots are 1-based, fspec.cc:4062;
// Rugra's coupled consumers are 0-based — outside the covered projection).

use rugra::address::Address;
use rugra::fspec::param_entry_flags;
use rugra::fspec::{ParamActive, ParamEntry, ParamListStandard, TypeClass};
use rugra::fspec::VarnodeData;
use rugra::space::{AddrSpace, AddressSpace, SpaceRegistry, SpaceType};

fn proc_space(name: &str, big_end: bool, size: u32, ws: u32, ind: i32, fl: u32) -> AddrSpace {
    AddrSpace::new_space(SpaceType::Processor, name, big_end, size, ws, ind, fl, 0, 0)
}

// canonical synthetic spaces: constant=0, other=1, unique=2, ram=3,
// register=4, stack=5 (same index plan as the C++ oracle fixture).
fn build_spaces(m: &mut SpaceRegistry) {
    m.insert_space(AddrSpace::new_constant_space(false)).unwrap();
    m.insert_space(AddrSpace::new_other_space()).unwrap();
    m.insert_space(AddrSpace::new_unique_space(2, 0, false)).unwrap();
    let ram_handle = proc_space("ram", false, 8, 1, 3, 1);
    m.insert_space(ram_handle.clone()).unwrap();
    m.insert_space(proc_space("register", false, 8, 1, 4, 1)).unwrap();
    m.insert_space(AddrSpace::new_spacebase_space(
        "stack", 5, 8, &ram_handle, 0, true, false,
    )).unwrap();
}

fn vd(v: &VarnodeData) -> String {
    format!("{}:0x{:x}/{}", v.space.name(), v.offset, v.size)
}

// staged-loader entry builder (the pub equivalent of the C++ fixture's
// direct field writes).
fn make_entry(
    grp: i32,
    spc: AddressSpace,
    base: u64,
    size: i32,
    min_size: i32,
    alignment: i32,
    flags: u32,
) -> ParamEntry {
    let mut e = ParamEntry::new(grp);
    *e.flags_mut() = flags;
    e.set_type_class(TypeClass::General);
    e.set_space(spc);
    e.set_base(base);
    e.set_sizes(size, min_size);
    e.set_alignment(alignment); // normalizes alignment == size to 0
    e
}

fn main() {
    println!("schema=1|fixture=FSPEC-PHASE0-1204|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    let mut registry = SpaceRegistry::new();
    build_spaces(&mut registry);
    let reg_handle = registry.get_space_by_name("register").unwrap();
    let stack_handle = registry.get_space_by_name("stack").unwrap();

    let coarse_reg = AddressSpace::Register;
    let coarse_stack = AddressSpace::Stack;
    let coarse_ram = AddressSpace::Ram;
    let coarse_unique = AddressSpace::Unique;
    let coarse_const = AddressSpace::Const;

    let mut model = ParamListStandard::new();
    let mut effects = Vec::new();
    model.parse_pentry(0, true, false, false, &mut effects, make_entry(
        0, coarse_reg, 0x30, 8, 4, 8, // alignment == size -> exclusion
        param_entry_flags::FORCE_LEFT_JUSTIFY,
    )).unwrap();
    model.parse_pentry(1, true, false, false, &mut effects, make_entry(
        1, coarse_reg, 0x38, 8, 4, 8,
        param_entry_flags::FORCE_LEFT_JUSTIFY | param_entry_flags::SMALLSIZE_ZEXT,
    )).unwrap();
    model.parse_pentry(2, true, false, false, &mut effects, make_entry(
        2, coarse_stack, 0, 64, 4, 8, // aligned, non-exclusion
        param_entry_flags::REVERSE_STACK,
    )).unwrap();
    model.parse_pentry(3, true, false, false, &mut effects, make_entry(
        3, coarse_ram, 0x2000, 16, 8, 16,
        param_entry_flags::FORCE_LEFT_JUSTIFY | param_entry_flags::SMALLSIZE_INTTYPE,
    )).unwrap();
    model.finalize_after_decode(0);
    let entries = model.get_entry().to_vec();

    // ---- case 1: find_entry per-space resolution ----
    println!("case=cross_space_find_entry");
    let pp: [(AddressSpace, u64, i32); 12] = [
        (coarse_reg, 0x30, 8),
        (coarse_reg, 0x30, 4),
        (coarse_reg, 0x34, 4),
        (coarse_reg, 0x30, 2),
        (coarse_reg, 0x48, 8),
        (coarse_stack, 0x8, 8),
        (coarse_stack, 0x10, 4),
        (coarse_stack, 0x100, 8),
        (coarse_ram, 0x2000, 16),
        (coarse_ram, 0x30, 8),
        (coarse_unique, 0x30, 8),
        (coarse_const, 0x2000, 8),
    ];
    for &(spc, off, sz) in pp.iter() {
        let mut slot: i32 = -1;
        let mut slot_size: i32 = -1;
        let ok = model.possible_param_with_slot(
            spc, Address::new(off), sz, &mut slot, &mut slot_size,
        );
        println!("  pp {}:{}/{} ok={} slot={} slotsize={}", spc.name(), hex(off), sz,
                 if ok { 1 } else { 0 }, slot, slot_size);
    }

    // ---- case 2: unjustified_container has no space filter ----
    println!("case=unjustified_container");
    let uc: [(AddressSpace, u64, i32); 5] = [
        (coarse_reg, 0x34, 4),
        (coarse_reg, 0x32, 4),
        (coarse_reg, 0x30, 4),
        (coarse_ram, 0x2008, 8),
        (coarse_stack, 0x2, 8),
    ];
    for &(spc, off, sz) in uc.iter() {
        let mut res = VarnodeData { space: AddressSpace::Ram, offset: 0, size: 0 };
        let hit = model.unjustified_container(Address::new(off), sz, &mut res);
        if hit {
            println!("  uc {}:{}/{} hit=1 res={}", spc.name(), hex(off), sz, vd(&res));
        } else {
            println!("  uc {}:{}/{} hit=0", spc.name(), hex(off), sz);
        }
    }

    // ---- case 3: assumed_extension has no space filter ----
    println!("case=assumed_extension");
    let ae: [(AddressSpace, u64, i32); 5] = [
        (coarse_reg, 0x38, 4),
        (coarse_ram, 0x2000, 8),
        (coarse_reg, 0x30, 4),
        (coarse_ram, 0x2004, 8),
        (coarse_reg, 0x38, 8),
    ];
    for &(spc, off, sz) in ae.iter() {
        let mut res = VarnodeData { space: AddressSpace::Ram, offset: 0, size: 0 };
        let op = model.assumed_extension(Address::new(off), sz, &mut res);
        if op != rugra::opcodes::OpCode::CPUI_COPY {
            println!("  ae {}:{}/{} op={} res={}", spc.name(), hex(off), sz, op.name(), vd(&res));
        } else {
            println!("  ae {}:{}/{} op=COPY", spc.name(), hex(off), sz);
        }
    }

    // ---- case 4: sort_trials uses the operator< ladder ----
    println!("case=sort_trials");
    {
        let mut active = ParamActive::new(false);
        active.register_trial(Address::with_space(&stack_handle, 0x10), 8);
        active.register_trial(Address::with_space(&reg_handle, 0x38), 4);
        active.register_trial(Address::with_space(&stack_handle, 0x0), 8);
        active.register_trial(Address::with_space(&reg_handle, 0x30), 8);
        active.register_trial(Address::with_space(&reg_handle, 0x34), 4);
        active.register_trial(Address::with_space(&reg_handle, 0x50), 8);
        active.get_trial_mut(0).set_entry(2, 0);
        active.get_trial_mut(1).set_entry(1, 0);
        active.get_trial_mut(2).set_entry(2, 0);
        active.get_trial_mut(3).set_entry(0, 0);
        active.get_trial_mut(4).set_entry(0, 4);
        active.sort_trials(&entries);
        for i in 0..active.get_num_trials() {
            let t = active.get_trial(i);
            let entry = t.get_entry_index();
            let grp = entry.map(|ix| entries[ix].get_group()).unwrap_or(-1);
            let off = entry.map(|_| t.get_offset()).unwrap_or(-1);
            let spc = t.get_address().get_space().map(|s| s.get_name()).unwrap_or_default();
            println!("  sorted[{}] addr={}:0x{:x} size={} grp={} off={}",
                     i, spc, t.get_address().as_u64(), t.get_size(), grp, off);
        }
    }

    // ---- case 5: split_hi/split_lo address, flags, slot ----
    println!("case=split_hi_lo");
    {
        let mut active = ParamActive::new(false);
        active.register_trial(Address::with_space(&reg_handle, 0x100), 12);
        {
            let t = active.get_trial_mut(0);
            t.mark_used();
            t.mark_active();
        }
        let (base_slot, base_addr) = {
            let t = active.get_trial(0);
            (t.get_slot(), t.get_address())
        };
        let hi = active.get_trial(0).split_hi(4);
        let lo = active.get_trial(0).split_lo(4);
        let name = |a: Address| a.get_space().map(|s| s.get_name()).unwrap_or_default();
        println!("  hi addr={}:0x{:x} size={} slot_delta={} used={} active={} checked={}",
                 name(hi.get_address()), hi.get_address().as_u64(), hi.get_size(),
                 hi.get_slot() - base_slot, hi.is_used() as i32, hi.is_active() as i32,
                 hi.is_checked() as i32);
        println!("  lo addr={}:0x{:x} size={} slot_delta={} used={} active={} checked={}",
                 name(lo.get_address()), lo.get_address().as_u64(), lo.get_size(),
                 lo.get_slot() - base_slot, lo.is_used() as i32, lo.is_active() as i32,
                 lo.is_checked() as i32);
        let lo8 = active.get_trial(0).split_lo(8);
        println!("  lo8 addr={}:0x{:x} size={} slot_delta={}",
                 name(lo8.get_address()), lo8.get_address().as_u64(), lo8.get_size(),
                 lo8.get_slot() - base_slot);
        let _ = base_addr;
    }

    // ---- case 6: split_trial rebuilds slots and keeps flags ----
    println!("case=split_trial");
    {
        let mut active = ParamActive::new(false);
        active.register_trial(Address::with_space(&reg_handle, 0x100), 12);
        active.register_trial(Address::with_space(&reg_handle, 0x200), 8);
        {
            let t = active.get_trial_mut(0);
            t.mark_used();
            t.mark_active();
        }
        let survivor_slot_before = active.get_trial(1).get_slot();
        let slotbase_before = active.get_slot_base();
        active.split_trial(0, 4);
        let name = |ix: usize| {
            active.get_trial(ix).get_address().get_space()
                .map(|s| s.get_name()).unwrap_or_default()
        };
        for i in 0..active.get_num_trials() {
            let t = active.get_trial(i);
            println!("  st[{}] addr={}:0x{:x} size={} used={} active={}",
                     i, name(i), t.get_address().as_u64(), t.get_size(),
                     t.is_used() as i32, t.is_active() as i32);
        }
        println!("  survivor_slot_delta={} slotbase_delta={} numtrials={}",
                 active.get_trial(2).get_slot() - survivor_slot_before,
                 active.get_slot_base() - slotbase_before,
                 active.get_num_trials());
    }
}

fn hex(v: u64) -> String {
    format!("0x{:x}", v)
}
